pkg/streamer/file.go

```diff
@@ -186,141 +187,86 @@ func fileSizeUpdated(path string, latestSize int64) (int, error) {
-				result <- watchResult{
-					updatePath: "",
-					err:        newCtx.Err(),
-				}
+	ticker := time.NewTicker(watcherInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			errCh <- terror.Annotate(ctx.Err(), "context meet error")
+			return
+		case <-ticker.C:
+			cmp, err := fileSizeUpdated(latestFilePath, latestFileSize)
+			if err != nil {
+				errCh <- terror.Annotatef(err, "latestFilePath=%s latestFileSize=%d", latestFilePath, latestFileSize)
 				return
-			case err2, ok := <-watcher2.Errors:
-				if !ok {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanClosed.Generate("errors", dir),
-					}
-				} else {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanRecvError.Delegate(err2, dir),
-					}
-				}
+			}
+			failpoint.Inject("CMPAlwaysReturn0", func() {
+				cmp = 0
+			})
+			switch {
+			case cmp < 0:
+				errCh <- terror.ErrRelayLogFileSizeSmaller.Generate(latestFilePath)
 				return
-			case event, ok := <-watcher2.Events:
-				if !ok {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanClosed.Generate("events", dir),
-					}
+			case cmp > 0:
+				updatePathCh <- latestFilePath
+				return
+			default:
+				// current watched file size have no change means that no new writes have been made
+				// our relay meta file will be updated immediately after receive the rotate event
+				// although we cannot ensure that the binlog filename in the meta is the next file after latestFile
+				// but if we return a different filename with latestFile, the outer logic (parseDirAsPossible)
+				// will find the right one
+				meta := &Meta{}
+				_, err = toml.DecodeFile(filepath.Join(dir, utils.MetaFilename), meta)
+				if err != nil {
+					errCh <- terror.Annotate(err, "decode relay meta toml file failed")
 					return
 				}
-				log.L().Debug("watcher receive event", zap.Reflect("event", event))
-				if event.IsDirEvent() {
-					log.L().Debug("skip watched event for directory", zap.Reflect("event", event))
-					continue
-				} else if !event.HasOps(watcher.Modify, watcher.Create) {
-					log.L().Debug("skip uninterested event", zap.Stringer("operation", event.Op), zap.String("path", event.Path))
-					continue
-				}
-				baseName := filepath.Base(event.Path)
-				if !binlog.VerifyFilename(baseName) {
-					log.L().Debug("skip watcher event for invalid relay log file", zap.Reflect("event", event))
-					continue // not valid binlog created, updated
-				}
-				result <- watchResult{
-					updatePath: event.Path,
-					err:        nil,
+				if meta.BinLogName != latestFile {
+					// we need check file size again, as the file may have been changed during our metafile check
+					cmp, err := fileSizeUpdated(latestFilePath, latestFileSize)
+					if err != nil {
+						errCh <- terror.Annotatef(err, "latestFilePath=%s latestFileSize=%d",
+							latestFilePath, latestFileSize)
```

Comment:
return here. And i'm requesting unit tests that "if relayLogUpdatedOrNewCreated returns, len(updatePathCh) + len(errCh) should be equal to 1"

Reply:
add in https://github.com/pingcap/dm/pull/1856/commits/1e4d300963b62604a66ecfc4e0d6aa9c5be2ef72

Reply:
RETURN HERE

Reply:
mybad... you are so attentive